haystack/tracing/tracer.py

```diff
@@ -0,0 +1,105 @@
+import abc
+import contextlib
+from typing import Dict, Any, ContextManager
+
+
+class Span(abc.ABC):
+    """Interface for an instrumented operation."""
+
+    @abc.abstractmethod
+    def set_tag(self, key: str, value: Any) -> None:
+        """Set a single tag on the span.
+
+        Note that the value will be serialized to a string, so it's best to use simple types like strings, numbers, or
+        booleans.
+
+        :param key: the name of the tag.
+        :param value: the value of the tag.
+        """
+        pass
+
+    def set_tags(self, tags: Dict[str, Any]) -> None:
+        """Set multiple tags on the span.
+
+        :param tags: a mapping of tag names to tag values.
+        """
+        for key, value in tags.items():
+            self.set_tag(key, value)
+
+
+class Tracer(abc.ABC):
+    """Interface for instrumenting code by creating and submitting spans."""
+
+    @abc.abstractmethod
+    @contextlib.contextmanager
+    def trace(self, operation_name: str, tags: Dict[str, Any] | None = None) -> ContextManager[Span]:
```

Comment:
compared to my initial plan I dropped support for linking parent spans because I think that should rather be done outside of `haystack` for example

```python
tracer = .... # e.g. datadog tracer

with tracer.trace(..., parent=Context(trace_id=..., span_id=...):
   pipeline.run(...)

```